File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {

        if (digits.empty()) {
            return {};
        }
        
        unordered_map<char, vector<string>> mp = {
            {'2', {"a", "b", "c"}},
            {'3', {"d", "e", "f"}},
            {'4', {"g", "h", "i"}},
            {'5', {"j", "k", "l"}},
            {'6', {"m", "n", "o"}},
            {'7', {"p", "q", "r", "s"}},
            {'8', {"t", "u", "v"}},
            {'9', {"w", "x", "y", "z"}}
        };

        vector<string> result =  mp[digits[0]];

        for(int i = 1;  i<digits.size(); i++){
                vector<string> current = mp[digits[i]];

                vector<string> temp;

                for(int j = 0; j<result.size(); j++){
                    for(int k = 0; k<current.size(); k++){
                        temp.push_back(result[j] + current[k]);
                    }
                }

                result = temp;
        }

        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {

    public void solve(
        int index,
        String digits,
        StringBuilder current,
        List<String> result,
        String[] mp
    ) {

        // Base case
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }

        // Get letters for current digit
        String letters = mp[digits.charAt(index) - '0'];

        // Try every letter
        for (char ch : letters.toCharArray()) {

            // Choose
            current.append(ch);

            // Explore
            solve(index + 1, digits, current, result, mp);

            // Undo
            current.deleteCharAt(current.length() - 1);
        }
    }

    public List<String> letterCombinations(String digits) {

        if (digits.length() == 0) {
            return new ArrayList<>();
        }

        String[] mp = {
            "", "", "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        List<String> result = new ArrayList<>();
        StringBuilder current = new StringBuilder();

        solve(0, digits, current, result, mp);

        return result;
    }
}